File: Provenance/Game Library/UI/PVConflictViewController.m
//
//  PVConflictViewController.m
//  Provenance
//
//  Created by James Addyman on 17/04/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

#import "PVConflictViewController.h"
#import "PVGameImporter.h"
#import "PVEmulatorConfiguration.h"

@interface PVConflictViewController ()

@property (nonatomic, strong, nonnull) PVGameImporter *gameImporter;
@property (nonatomic, strong, nullable) NSArray<NSString*> *conflictedFiles;

@end

@implementation PVConflictViewController

- (instancetype)initWithGameImporter:(PVGameImporter * __nonnull)gameImporter
{
    if ((self = [super initWithStyle:UITableViewStylePlain]))
    {
        self.gameImporter = gameImporter;
    }
    
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];

#if TARGET_OS_TV
    self.splitViewController.title = @"Solve Conflicts";
#else
    self.title = @"Solve Conflicts";

    if (![[self conflictedFiles] count]) {
        self.tableView.separatorColor = [UIColor clearColor];
    }
#endif
    [self updateConflictedFiles];
}

- (void)viewWillAppear:(BOOL)animated;
{
    [super viewWillAppear:animated];
    
    if (!self.navigationController || self.navigationController.viewControllers.count<=1) {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                                                               target:self
                                                                                               action:@selector(dismiss)];
    }
}

- (void)dismiss;
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)updateConflictedFiles
{
    NSMutableArray *tempConflictedFiles = [NSMutableArray array];
    NSArray *filesInConflictsFolder = [self.gameImporter conflictedFiles];
    for (NSString *file in filesInConflictsFolder)
    {
        NSString *extension = [file pathExtension];
        if ([[[PVEmulatorConfiguration sharedInstance] systemIdentifiersForFileExtension:[extension lowercaseString]] count] > 1)
        {
            [tempConflictedFiles addObject:file];
        }
    }
    
    self.conflictedFiles = [tempConflictedFiles copy];
}

- (NSString *)documentsPath
{
#if TARGET_OS_TV
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
#else
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
#endif
    NSString *documentsDirectoryPath = [paths objectAtIndex:0];
    
    return documentsDirectoryPath;
}

- (NSString *)conflictsPath
{
#if TARGET_OS_TV
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
#else
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
#endif
    NSString *documentsDirectoryPath = [paths objectAtIndex:0];
    
    return [documentsDirectoryPath stringByAppendingPathComponent:@"conflicts"];
}

#if TARGET_OS_TV
- (BOOL)tableView:(UITableView *)tableView canFocusRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (![self.conflictedFiles count])
    {
        if ([indexPath row] == 2) {
            return YES;
        }
        
        return NO;
    }

    return YES;
}
#endif

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if ([self.conflictedFiles count])
    {
        return [self.conflictedFiles count];
    }

    return 3;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (![self.conflictedFiles count])
    {

        UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"EmptyCell"];
        if (!cell)
        {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"Cell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
        }

        cell.textLabel.textAlignment = NSTextAlignmentCenter;

        if ([indexPath row] == 0 || [indexPath row] == 1)
        {
            cell.textLabel.text = @"";
        }
        else
        {
            cell.textLabel.text = @"No Conflicts...";
        }

        return cell;
    }

    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"Cell"];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"Cell"];
    }
    NSString *file = self.conflictedFiles[[indexPath row]];
    NSString *name = [[file lastPathComponent] stringByReplacingOccurrencesOfString:[@"." stringByAppendingString:[file pathExtension]] withString:@""];
    
    [[cell textLabel] setText:name];
    [cell setAccessoryType:UITableViewCellAccessoryDisclosureIndicator];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [self.tableView deselectRowAtIndexPath:[self.tableView indexPathForSelectedRow] animated:YES];

    if (![self.conflictedFiles count])
    {
        return;
    }

    NSString *path = self.conflictedFiles[[indexPath row]];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Choose a System"
                                                                             message:nil
                                                                      preferredStyle:UIAlertControllerStyleActionSheet];
    [[alertController popoverPresentationController] setSourceView:self.view];
    [[alertController popoverPresentationController] setSourceRect:[self.tableView rectForRowAtIndexPath:indexPath]];
    
    for (NSString *systemID in [[PVEmulatorConfiguration sharedInstance] availableSystemIdentifiers])
    {
        NSArray *supportedExtensions = [[PVEmulatorConfiguration sharedInstance] fileExtensionsForSystemIdentifier:systemID];
        if ([supportedExtensions containsObject:[path pathExtension]])
        {
            NSString *name = [[PVEmulatorConfiguration sharedInstance] shortNameForSystemIdentifier:systemID];
            [alertController addAction:[UIAlertAction actionWithTitle:name style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                [self.gameImporter resolveConflictsWithSolutions:@{path: systemID}];
                // This update crashes since we remove for me on aTV.
//                [self.tableView beginUpdates];
//                [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationTop];
                [self updateConflictedFiles];
                [self.tableView reloadData];
//                [self.tableView endUpdates];
            }]];
        }
    }
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:NULL]];
    [self presentViewController:alertController animated:YES completion:NULL];
}

@end

File: Provenance/Game Library/UI/PVConflictViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6632 - https://objectivec2swift.com/
//
//  PVConflictViewController.swift
//  Provenance
//
//  Created by James Addyman on 17/04/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

import PVLibrary
import PVSupport
import UIKit
import RxSwift
import RxDataSources

final class PVConflictViewController: UITableViewController {
    let conflictsController: ConflictsController
    private let disposeBag = DisposeBag()

    init(conflictsController: ConflictsController) {
        self.conflictsController = conflictsController
        super.init(style: .plain)
    }

    required init?(coder _: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    enum Row {
        case conflict(ConflictsController.Conflict)
        case empty(title: String)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Solve Conflicts"
        #if !os(tvOS)
            let currentTableview = tableView!
            tableView = UITableView(frame: currentTableview.frame, style: currentTableview.style)
            tableView.separatorColor = UIColor.clear
        #endif

		tableView.delegate = nil
		tableView.dataSource = nil

        let cellIdentifier = "Cell"
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)

        let rows = conflictsController.conflicts
            .map({ conflicts -> [Row] in
                if conflicts.isEmpty {
                    return [.empty(title: "No Conflicts…")]
                } else {
                    return conflicts.map { .conflict($0) }
                }
            })

        rows.bind(to: tableView.rx.items(cellIdentifier: cellIdentifier, cellType: UITableViewCell.self)) { _, row, cell in
            switch row {
            case .conflict(let conflict):
                cell.editingAccessoryType = .checkmark
                cell.textLabel?.text = conflict.path    .lastPathComponent
                cell.accessoryType = .disclosureIndicator
            case .empty(let title):
                cell.textLabel?.text = title
                cell.textLabel?.textAlignment = .center
                cell.accessoryType = .none
                cell.isUserInteractionEnabled = false
            }
        }
        .disposed(by: disposeBag)

        tableView.rx.itemDeleted
            .do(onNext: {
                self.tableView.deselectRow(at: $0, animated: true)
            })
                .compactMap({ indexPath -> (ConflictsController.Conflict, IndexPath)? in
                    let row: Row = try self.tableView.rx.model(at: indexPath)
                    switch row {
                    case .conflict(let conflict):
                        return (conflict, indexPath)
                    case .empty:
                        return nil
                    }
                })
                .bind(onNext: { conflict, indexPath in
                    self.conflictsController.deleteConflict(path: conflict.path)
                    self.tableView.reloadData()
                })
            .disposed(by: disposeBag)

        tableView.rx.itemSelected
            .do(onNext: { self.tableView.deselectRow(at: $0, animated: true) })
            .compactMap({ indexPath -> (ConflictsController.Conflict, IndexPath)? in
                let row: Row = try self.tableView.rx.model(at: indexPath)
                switch row {
                case .conflict(let conflict):
                    return (conflict, indexPath)
                case .empty:
                    return nil
                }
            })
            .bind(onNext: { conflict, indexPath in
                let showsUnsupportedSystems = PVSettingsModel.shared.debugOptions.unsupportedCores
                let alertController = UIAlertController(title: "Choose a System", message: nil, preferredStyle: .actionSheet)
                alertController.popoverPresentationController?.sourceView = self.view
                alertController.popoverPresentationController?.sourceRect = self.tableView.rectForRow(at: indexPath)
                conflict.candidates
                    .filter { $0.supported || showsUnsupportedSystems }
                    .sorted(by: { sys1,sys2 in sys1.identifier < sys2.identifier })
                    .forEach { system in
                        alertController.addAction(.init(title: system.name, style: .default, handler: { _ in
                            self.conflictsController.resolveConflicts(withSolutions: [conflict.path: system])
                        }))
                    }
                alertController.addAction(.init(title: NSLocalizedString("Delete", comment: "Delete file"), style: .destructive, handler: { _ in
                    self.conflictsController.deleteConflict(path: conflict.path)
                }))
                alertController.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: "Cancel"), style: .cancel, handler: nil))
                self.present(alertController, animated: true) { () -> Void in
                    self.tableView.reloadData()
                }
            })
            .disposed(by: disposeBag)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        #if !os(tvOS)
        if navigationController == nil || navigationController!.viewControllers.count <= 1 {
            navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(showEditing))
            navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(dismissMe))
        }
        #endif
    }

    @objc func dismissMe() {
        dismiss(animated: true) { () -> Void in }
    }
}

extension PVConflictViewController {

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    override func tableView(_ tableView: UITableView, shouldIndentWhileEditingRowAt indexPath: IndexPath) -> Bool {
        return true
    }

    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return .delete
    }

    @objc private func toggleEditing() {
        tableView.setEditing(!tableView.isEditing, animated: true) // Set opposite value of current editing status
        navigationItem.rightBarButtonItem?.title = tableView.isEditing ? "Done" : "Edit" // Set title depending on the editing status
    }

    @objc func showEditing(sender: UIBarButtonItem) {
       if self.tableView.isEditing {
           self.tableView.isEditing = false
           self.navigationItem.rightBarButtonItem?.title = "Done"
       } else {
           self.tableView.isEditing = true
           self.navigationItem.rightBarButtonItem?.title = "Edit"
       }
   }

    override func setEditing(_ editing: Bool, animated: Bool) {
        super.setEditing(editing, animated: animated)

        // Toggle table view editing.
         tableView.setEditing(editing, animated: true)
    }
}
